crates/sncast/src/starknet_commands/declare.rs

```diff
@@ -78,24 +73,39 @@ pub async fn declare(
         .class_hash()
         .map_err(anyhow::Error::from)?;
 
-    let FeeSettings {
-        max_gas,
-        max_gas_unit_price,
-    } = fee_settings;
     let declaration = account.declare_v3(
         Arc::new(contract_definition.flatten().map_err(anyhow::Error::from)?),
         casm_class_hash,
     );
 
+    let fee_settings = if declare.fee_args.max_fee.is_some() {
+        let fee_estimate = declaration
+            .estimate_fee()
+            .await
+            .expect("Failed to estimate fee");
+        declare.fee_args.try_into_fee_settings(&Some(fee_estimate))
+    } else {
+        declare.fee_args.try_into_fee_settings(&None)
+    };
+
+    let FeeSettings {
+        l1_gas,
+        l1_gas_price,
+        l2_gas,
+        l2_gas_price,
+        l1_data_gas,
+        l1_data_gas_price,
+    } = fee_settings.expect("Failed to convert to fee settings");
+
+    let declaration = apply_optional(declaration, l1_gas, DeclarationV3::l1_gas);
+    let declaration = apply_optional(declaration, l1_gas_price, DeclarationV3::l1_gas_price);
+    let declaration = apply_optional(declaration, l2_gas, DeclarationV3::l2_gas);
+    let declaration = apply_optional(declaration, l2_gas_price, DeclarationV3::l2_gas_price);
+    let declaration = apply_optional(declaration, l1_data_gas, DeclarationV3::l1_data_gas);
```

Comment:
This logic is repeated in multiple places, can't this be extracted to some generic helper function?

Reply:
Or a macro if that's needed

Reply:
It can, sure - added a macro for that.